src/kakarot/library.cairo

```diff
@@ -269,9 +282,118 @@ namespace Kakarot {
+            let (bytecode_len, bytecode) = IAccount.bytecode(
+                contract_address=starknet_contract_address
+            );
+            let summary = execute(
+                starknet_contract_address,
+                to,
+                bytecode_len,
+                bytecode,
+                data_len,
+                data,
+                value,
+                gas_limit,
+                gas_price,
+            );
+            return (summary.return_data_len, summary.return_data);
+        }
+    }
+
+    // @notice Assert that the calling contract is a Kakarot account, ie. that it has been
+    //         deployed by kakarot. This is currently required to make sure that evm and
+    //         starknet addresses remain consistents along an execution
+    // @dev Raise if the declared corresponding evm address (retrieved with get_evm_address)
+    //      does not lead to the actual caller address
+    func assert_caller_is_kakarot_account{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }() {
+        alloc_locals;
+        let (local actual_starknet_address) = get_caller_address();
+        let (local declared_evm_address) = IAccount.get_evm_address(actual_starknet_address);
+        let (local computed_starknet_address) = Accounts.compute_starknet_address(
+            declared_evm_address
+        );
+
+        with_attr error_message("Kakarot: caller contract is not a Kakarot Account") {
+            assert computed_starknet_address = actual_starknet_address;
+        }
+
+        return ();
+    }
+
+    // @notice The eth_send_transaction function as described in the RPC spec,
+    //         see https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_sendtransaction
+    // @dev "from" parameter is taken from the get_caller_address syscall
+    // @dev "nonce" parameter is taken from the caller account
+    // @param to The address the transaction is directed to.
+    // @param gas_limit Integer of the gas provided for the transaction execution
+    // @param gas_price Integer of the gas price used for each paid gas
+    // @param value Integer of the value sent with this transaction
+    // @param data_len The length of the data
+    // @param data Hash of the method signature and encoded parameters. For details see Ethereum Contract ABI in the Solidity documentation
+    // @return return_data_len The length of the returned bytes
+    // @return return_data The returned bytes array
+    func eth_send_transaction{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(to: felt, gas_limit: felt, gas_price: felt, value: felt, data_len: felt, data: felt*) -> (
+        return_data_len: felt, return_data: felt*
+    ) {
+        assert_caller_is_kakarot_account();
```

Comment:
can you explain the logic here?

as I understand, 
- library function `Library.eth_call` can be called only as @view in kakarot contract -> won't trigger a tx
- `eth_send_transaction`  can be called as @external in kakarot contract. And here, we assert the caller is the Kakarot contract, and then call Library.eth_call -> that means we are in the context of send_transaction and as such there is a tx being posted?

Reply:
> we assert the caller is the Kakarot contract

we don't assert that the caller is **the kakarot contract**, we assert that the calling starknet contract is **a kakarot account**, ie an EOA/CA deployed through kakarot. The problem here is that any starknet contract can implement `get_evm_address` and returns whatever they want. Without this check, such a contract could trigger a transaction to Kakarot pretending to be a given EVM address